src/BTree.mo

```diff
@@ -0,0 +1,110 @@
+/// Imperative sequences as B-Trees.
+
+import A "Array";
+import I "Iter";
+import List "List";
+import Option "Option";
+import Order "Order";
+import P "Prelude";
+import Prim "mo:⛔";
+
+module {
+
+  /// Constants we use to shape the tree.
+  /// See https://en.wikipedia.org/wiki/B-tree#Definition
+  module Constants {
+    let MAX_CHILDREN = 4;
+  };
+
+  public type Compare<K> = {
+    compare : (K, K) -> Order.Order
+  };
+
+  public type Data<K, V> = [(K, V)];
+
+  public type Index<K, V> = {
+    data : Data<K, V>;
+    trees : [Tree<K, V>];
+  };
+
+  public type Tree<K, V> = {
+    #index : Index<K, V>;
+    #data : Data<K, V>;
+  };
+
+  func find_data<K, V>(data : Data<K, V>, find_k : K, c : Compare<K>) : ?V {
+    for ((k, v) in data.vals()) {
+      if (c.compare(k, find_k) == #equal) { return ?v };
+    };
+    return null
+  };
+
+  func find<K, V>(t : Tree<K, V>, k : K, c : Compare<K>) : ?V {
+    switch t {
+      case (#data(d)) { return find_data<K, V>(d, k, c) };
+      case (#index(i)) {
+        for (j in I.range(0, i.data.size())) {
+          switch (c.compare(k, i.data[j].0)) {
+            case (#equal) { return ?i.data[j].1 };
+            case (#less) { return find<K, V>(i.trees[j], k, c) };
+            case _ { }
+          }
+        };
+        find<K, V>(i.trees[i.data.size()], k, c)
+      };
+    };
+  };
+
+  /// Check that a B-Tree instance observes invariants of B-Trees.
+  /// Invariants ensure performance is what we expect.
+  /// For testing and debugging.
+  public module Check {
+
+    type CompareOp<K> = {
+      compare : (?K, ?K) -> Order.Order
+    };
+
+    func compareOp<K>(c : Compare<K>) : CompareOp<K> = {
+      compare = func (k1 : ?K, k2 : ?K) : Order.Order {
+        switch (k1, k2) {
+        case (null, null) { assert false; loop {} };
+        case (null, _) #less;
+        case (_, null) #greater;
+        case (?k1, ?k2) c.compare(k1, k2)
+        }
+      }
+    };
+
+    public func check<K, V>(c : Compare<K>, t : Tree<K, V>) {
+      rec(null, compareOp(c), t, null)
+    };
+
+    func rec<K, V>(lower : ?K, c : CompareOp<K>, t : Tree<K, V>, upper : ?K) {
+      switch t {
+        case (#data(d)) { data(lower, c, d, upper) };
+        case (#index(i)) { index(lower, c, i, upper) };
+      }
+    };
+
+    func data<K, V>(lower : ?K, c : CompareOp<K>, d : Data<K, V>, upper : ?K) {
+      var prev_k : ?K = null;
+      for ((k, _) in d.vals()) {
+        assert (c.compare(prev_k, ?k) != #greater);
+        assert (c.compare(lower, ?k) != #greater);
+        assert (c.compare(?k, upper) != #greater);
+        prev_k := ?k;
+      }
+    };
+
+    func index<K, V>(lower : ?K, c : CompareOp<K>, i : Index<K, V>, upper : ?K) {
+      assert (i.data.size() + 1 == i.trees.size());
```

Comment:
Why does this keep the invariant? I thought `i.data` was just `[(K, V)]`, whereas `i.trees` could be any size array specified by the BTree child size

Reply:
Perhaps I don't understand the question entirely, but the arrays of (sub)`trees` and keys (in `data`) are highly related under my current understanding of B-Trees, and the way they work in the rust implementation.

From that work, I get that this property holds (related to `assert (i.data.size() + 1 == i.trees.size());`)

> For each pair of consecutive subtrees (`t1`, `t2`), there is a key `k12` (and value `v12`) such that `keysOf(t1) < k12 < keysOf(t2)`

(abusing notation to compare sets of keys with a single key; what I mean is every key in each set observes the relevant comparison operator holding.)

I found it very helpful to look at pictures that people drew, and draw some myself.  For instance, [the one on the wikipedia page shows this pattern where keys are shown between consecutive subtrees.](https://en.wikipedia.org/wiki/B-tree#/media/File:B-tree.svg).

Reply:
PS: I strongly suspect that this invariant code is wrong.  It's a baby step that will need to be revised, probably a few times.  Thanks for taking a look!

Reply:
I'm cribbing the type structure from [this Rust implementation from `std`, also linked above in another thread](https://github.com/rust-lang/rust/blob/8c1cc82a82ebfbea731258063115992b2dd4adee/library/alloc/src/collections/btree/node.rs#L9).

Reply:
What makes the B-Tree "tricky" as a simple type def, IMO is the combination of array structure and tree structure, and how it mixes them together in an efficient implementation that uses arrays to organize things of the same type.

In a "usual" internal node, there is a key-value between each pair of consecutive sub-trees, giving a key that separates these sub-trees' keys.

But there is no key-value pair on either end of the sub-tree list, so in total, there is one fewer key-value pairs than sub-trees.

The Rust implementation linked above does this, based on me reading the comments (not the code itself).

Reply:
Got it, I understand completely now! Thanks for explaining.